### PR TITLE
Doing [filter_attack/weapon]type= pick type who doing more damage only

### PR DESCRIPTION
**Pull request by newfrenchy83**
Fix (https://github.com/wesnoth/wesnoth/issues/8585)

### CODE REVIEW

**Comment by soliton-** [comment]
There is handling of a `double_type` attribute added here with no explanation. Please either remove or explain it.

File: data/test/scenarios/wml_tests/ScenarioWML/EventWML/events-test_filter_attack_type.cfg [hunk 1]
```diff
@@ -84,8 +86,46 @@
...
+# An event not triggers when Alice attacks during side 1's turn because type=arcane can't be detected.
+#####
+{GENERIC_UNIT_TEST event_test_filter_attack_type_fail (
+    {FILTER_TYPE 50}
+
+    [event]
+        name=turn 2
+        {RETURN ({VARIABLE_CONDITIONAL triggers equals 0})}
+    [/event]
+)}
```
Comment (stevecotton) [inline data/test/scenarios/wml_tests/ScenarioWML/EventWML/events-test_filter_attack_type.cfg:131]: If `FILTER_TYPE` is only for use in this file then there should be an `#undef` at the end.

File: data/test/scenarios/wml_tests/ScenarioWML/EventWML/events-test_filter_attack_type.cfg [hunk 2]
```diff
@@ -84,8 +86,46 @@
...
+# API(s) being tested: [event][filter_attack]type=
+##
+# Actions:
+# Give Alice an ability that adds a damage special with addition of arcnae type to all of his weapons.
+# Give Bob resistance to -100% to arcane
+# Define events that use filter_attack matching Alice's arcane type.
+# Have Alice attack Bob.
+##
+# Expected end state:
+# An event triggers when Alice attacks during side 1's turn because type=arcane detected.
```
Comment (stevecotton) [inline data/test/scenarios/wml_tests/ScenarioWML/EventWML/events-test_filter_attack_type.cfg:101]: An event triggers when Alice attacks because the attack is calculated using arcane damage

File: data/test/scenarios/wml_tests/ScenarioWML/EventWML/events-test_filter_attack_type.cfg [hunk 3]
```diff
@@ -84,8 +86,46 @@
...
+# API(s) being tested: [event][filter_attack]type=
+##
+# Actions:
+# Give Alice an ability that adds a damage special with addition of arcane type to all of his weapons.
+# Give Bob resistance to 50% to arcane
+# Define events that use filter_attack matching Alice's arcane type.
+# Have Alice attack Bob.
+##
+# Expected end state:
+# An event not triggers when Alice attacks during side 1's turn because type=arcane can't be detected.
```
Comment (stevecotton) [inline data/test/scenarios/wml_tests/ScenarioWML/EventWML/events-test_filter_attack_type.cfg:122]: The event does not trigger when Alice attacks, because the result is calculated using blade or pierce as the damage type

File: data/test/scenarios/wml_tests/ScenarioWML/EventWML/events-test_filter_attack_type.cfg (hunk 1 above)
Comment (newfrenchy83) [reply data/test/scenarios/wml_tests/ScenarioWML/EventWML/events-test_filter_attack_type.cfg:131]: done

File: data/test/scenarios/wml_tests/ScenarioWML/EventWML/events-test_filter_attack_type.cfg (hunk 3 above)
Comment (newfrenchy83) [reply data/test/scenarios/wml_tests/ScenarioWML/EventWML/events-test_filter_attack_type.cfg:122]: yes,

File: data/test/scenarios/wml_tests/ScenarioWML/EventWML/events-test_filter_attack_type.cfg (hunk 2 above)
Comment (newfrenchy83) [reply data/test/scenarios/wml_tests/ScenarioWML/EventWML/events-test_filter_attack_type.cfg:101]: yes

File: src/units/abilities.cpp [hunk 4]
```diff
@@ -1274,6 +1274,24 @@ std::set<std::string> attack_type::alternative_damage_types() const
...
+
+	unit_ability_list resistance_list;
+	std::string type_damage = damages_type.first;
+	if(other_){
+		resistance_list = (*other_).get_abilities_weapons("resistance", other_loc_, other_attack_, shared_from_this());
+		int first_value = (*other_).resistance_value(resistance_list, damages_type.first);
+		int secondary_value = (*other_).resistance_value(resistance_list, damages_type.second);
+		type_damage = secondary_value > first_value ? damages_type.second : damages_type.first;
+	}
+	return type_damage;
```
Comment (stevecotton) [inline src/units/abilities.cpp:1292]: I assume this is adding another copy of existing logic.

Could filter_damage_type be implemented by calling an existing function?
Could an existing function be simplified by calling filter_damage_type instead (in which case, I'd suggest renaming filter_damage_type to effective_damage_type)?

File: src/units/abilities.cpp [hunk 5]
```diff
@@ -1274,6 +1274,24 @@ std::set<std::string> attack_type::alternative_damage_types() const
 	return damage_types;
 }
 
+std::string attack_type::filter_damage_type() const
+{
+	std::pair<std::string, std::string> damages_type = damage_type();
+	if((damages_type.second).empty()){
```
[the remark this answers is not in the document]
Comment (newfrenchy83) [reply src/units/abilities.cpp:1280]: 'there is no alternative_type the rest of the code is useless and you might as well immediately return the value type|replacement_type without comparing the resistances

[the remark this answers is not in the document]
Comment (soliton-) [reply src/units/abilities.cpp:1280]: Sure, and if there is an alternative_type you're done just as well..? Not sure what you're getting at.

[the remark this answers is not in the document]
Comment (newfrenchy83) [reply src/units/abilities.cpp:1280]: What I'm trying to say is that if alternative_type is not specified here, then damage_type.first must be returned immediately without going through the resistance calculations

File: src/units/abilities.cpp (hunk 4 above)
Comment (newfrenchy83) [reply src/units/abilities.cpp:1307]: done

File: wml_test_schedule [hunk 6]
```diff
@@ -160,6 +160,7 @@
 0 event_test_filter_attack
 0 event_test_filter_attack_no_defense
 0 event_test_filter_attack_type
+0 event_test_filter_attack_type_fail
```
Comment (stevecotton) [inline wml_test_schedule:163]: The suffix `_fail` should be reserved for tests which return a non-zero status. Please rename it.

File: src/units/attack_type.hpp [hunk 7]
```diff
@@ -99,6 +99,8 @@ class attack_type : public std::enable_shared_from_this<attack_type>
 	std::pair<std::string, std::string> damage_type() const;
 	/** @return A list of alternative_type damage types. */
 	std::set<std::string> alternative_damage_types() const;
+	/** @return A modified damage type or a alternative_type based on highest damage */
```
Comment (stevecotton) [inline src/units/attack_type.hpp:102]: I'm not sure what to suggest yet, but we need better documentation - there's too many similarly-named functions here where neither the name nor the documentation answers the question "which one should I use".

File: wml_test_schedule (hunk 6 above)
Comment (newfrenchy83) [reply wml_test_schedule:163]: Because bug in has_attack filter, i sould probably close this Pr and same remove replacement_type  for 1.20

File: src/units/abilities.cpp [hunk 8]
```diff
@@ -1239,24 +1239,31 @@ std::string attack_type::select_damage_type(const unit_ability_list& damage_type
 /**
  * Returns the type of damage inflicted.
```
Comment (stevecotton) [inline src/units/abilities.cpp:1240]: This documentation is now out of date

File: wml_test_schedule (hunk 6 above)
Comment (newfrenchy83) [reply wml_test_schedule]: name changed

File: src/units/abilities.cpp (hunk 8 above)
Comment (newfrenchy83) [reply src/units/abilities.cpp]: fixed

File: src/units/attack_type.hpp (hunk 7 above)
Comment (newfrenchy83) [reply src/units/attack_type.hpp]: documentation changed

**Comment by newfrenchy83** [comment]
@stevecotton could youreview anew this PR? Thanks.

**Comment by stevecotton** [comment]
I know that I need to spend time on this, but intend to only work on the infinite recursion bug until that's fixed.

**Comment by newfrenchy83** [comment]
@Pentarctagon there are 3 of my PRs left to merge into the master branch that I rank by priority: this one, https://github.com/wesnoth/wesnoth/pull/9476 and https://github.com/wesnoth/wesnoth/pull/9079
https://github.com/wesnoth/wesnoth/pull/9476 concerns the replacement of the truncation of values ​​by a rounding to the nearest integer and https://github.com/wesnoth/wesnoth/pull/9476 the use of a [filter][filter_adjacent_location] to replace [filter_adjacent_location] in the abilities that have become inoperative

**Comment by newfrenchy83** [comment]
@stevecotton @Pentarctagon someone culd review this PR, please?

**Comment by Pentarctagon** [comment]
It'd be good to pull in the recent clang-tidy commits to make sure your PRs run with those added checks.

**Comment by newfrenchy83** [comment]
> It'd be good to pull in the recent clang-tidy commits to make sure your PRs run with those added checks.

I rebased and force pushed this PR today.

**Comment by Pentarctagon** [comment]
There should be tests for the new `original_type` attribute.

**Comment by newfrenchy83** [comment]
> There should be tests for the new `original_type` attribute.

@Pentarctagon test added in https://github.com/wesnoth/wesnoth/pull/8936/commits/0eb694e1396a8e17f02ee06476ecf25718de261d commit

**Comment by Pentarctagon** [comment]
Seems the tests are failing though.

**Comment by newfrenchy83** [comment]
> Seems the tests are failing though.

it's work now.

File: src/formula/callable_objects.cpp [hunk 9]
```diff
@@ -87,8 +87,10 @@ variant attack_type_callable::get_value(const std::string& key) const
 		return variant(att_->id());
 	} else if(key == "description") {
 		return variant(att_->name());
-	} else if(key == "type") {
+	} else if(key == "original_type") {
 		return variant(att_->type());
+	} else if(key == "type") {
+		return variant(att_->effective_damage_type().first);
```
Comment (Pentarctagon) [inline src/formula/callable_objects.cpp:93]: @CelticMinstrel any concerns about changing what `type` refers to here?

File: src/formula/callable_objects.cpp [hunk 10]
```diff
@@ -173,6 +176,10 @@ int attack_type_callable::do_compare(const formula_callable* callable) const
 		return att_->type().compare(att_callable->att_->type());
 	}
 
+	if(att_->effective_damage_type().first != att_callable->att_->effective_damage_type().first) {
+		return att_->effective_damage_type().first.compare(att_callable->att_->effective_damage_type().first);
+	}
```
Comment (Pentarctagon) [inline src/formula/callable_objects.cpp:181]: What is the purpose of this check?

**Comment by Pentarctagon** [comment]
I assume you've tested that the UI-related changes are working correctly?

File: src/formula/callable_objects.cpp (hunk 10 above)
Comment (newfrenchy83) [reply src/formula/callable_objects.cpp:181]: @Pentarctagon @CelticMinstrel if checking the type actually used in a formula is problematic, I can remove it, I just wanted to harmonize the language with [filter_weapon]

**Comment by newfrenchy83** [comment]
> I assume you've tested that the UI-related changes are working correctly?

Yes, I did the test, and everything works as I want, the pre-combat window only shows the effective type, while the unit report bar shows the icons of the damage types available for each weapon

File: src/formula/callable_objects.cpp (hunk 9 above)
Comment (CelticMinstrel) [reply src/formula/callable_objects.cpp:93]: It definitely looks concerning. I have no idea what `effective_damage_type()` is doing but it's not the same as what `type()` returns, so the correct answer is probably to leave `"type"` alone here and add an `"effective_type"` that uses the new function.

Comment (newfrenchy83) [reply src/formula/callable_objects.cpp:93]: effective_damage_type() return the type used in attack same if modified by use of [damage_type], but i will apply your suggest

Comment (newfrenchy83) [reply src/formula/callable_objects.cpp:93]: @CelticMinstrel the problem is that "type" in attack_type refers to the type used, including after modification by [damage_type]; it therefore seems logical to me to extend this to the formula and use "original_type" to refer to the type not modified by [damage_type], either I apply your suggestion, and in this case it implies a regression for [filter_attack], or I give up original type, and only filter or checked in the formula the type actually used, but not say 'type' refers to the damage type actually used in the filter but only refers to the original type in the formula

Comment (CelticMinstrel) [reply src/formula/callable_objects.cpp:93]: I don't really understand what you're saying. The `type()` function just returns the string stored in the class, which presumably is the type set from WML in the attack definition. Changing it to account for `[damage_type]` specials seems like it would be more confusing than useful, as that leads to infinite recursion in some circumstances – although I recall measures being taken against that, breaking the recursion still means the outcome is not well-defined. And changing the meaning of existing formulas also doesn't sound like a great idea.

But we can discuss this. There may be a third option too, or even fourth or fifth options.

> either I apply your suggestion, and in this case it implies a regression for [filter_attack]

Please explain the regression that would occur if you apply my suggestion.

> or I give up original type, and only filter or checked in the formula the type actually used, but not say 'type' refers to the damage type actually used in the filter but only refers to the original type in the formula

I'm really trying hard to parse this run-on sentence. From what I can tell, you are suggesting to just undo this part of the diff and leave it how it previously was, and also adding that a clarification should be made to the documentation?

I think there _is_ some benefit to having both versions of `type`, so that doesn't seem like a great option. However, I think it is not at all obvious whether it makes more sense for `type` to be the original type or the effective type. That's why I think it's better to leave `type` unchanged.

-----

By the way, as an aside: whatever we ultimately decide on here, the same change should also be applied in the Lua API.

Comment (newfrenchy83) [reply src/formula/callable_objects.cpp:93]: i will remove all modifications in formula in regard of infinite recursion issue. For LUA API, i cannot apply change, i already try for attack_alignment attribute and it never worked, because data of units(self or opponent) can't be added in attack attribute in lua, and it will the same for [damage_type] weapon

Comment (newfrenchy83) [reply src/formula/callable_objects.cpp]: @CelticMinstrel [filter_attack]type= already takes into account the modifications by [damage_type], except that it integrates at the same time the original type (or replace by [damage_type]replacement_type=) and the alternative type ([damage_type]alternative_type=), here, I want that only the type of damage which does the most damage to the opponent is taken into account and appears in the pre-combat window, as well as for the check by [filter_attack/weapon]

File: src/formula/callable_objects.cpp (hunk 10 above)
Comment (newfrenchy83) [reply src/formula/callable_objects.cpp]: formula modification removed

File: src/formula/callable_objects.cpp (hunk 9 above)
Comment (CelticMinstrel) [reply src/formula/callable_objects.cpp]: So in summary, you're saying that your goal is to make the WFL API behave more like as the filter API with respect to the meaning of `type`? Keeping them both the same does make some sense, but… you also seem to be saying you want it to be _different_ from how `[filter_attack]type=` works. I find that difficult to reconcile. It seems like there might be three different possible versions of `type` here?

File: src/formula/callable_objects.cpp (hunk 10 above)
Comment (Pentarctagon) [reply src/formula/callable_objects.cpp]: And yet my question remains unanswered.

File: src/formula/callable_objects.cpp (hunk 9 above)
Comment (newfrenchy83) [reply src/formula/callable_objects.cpp]: @CelticMinstrel  I hesitate because of the infinite recursion if the formula is used from a [damage_type] affecting the checked attack, but if I modify the code of the formula, it is good for the attributes of filter_weapon and the formula to be identical. So, do I modify formula or not?

File: src/formula/callable_objects.cpp (hunk 10 above)
Comment (newfrenchy83) [reply src/formula/callable_objects.cpp]: @Pentarctagon This check is the code to check the type of damage actually inflicted while the old code for type() is reassigned to 'original_type', in short if you use 'type' in a formula in [filter_attack/weapon], it is the type of damage that will actually be used that will be detected, a bit like if you use [filter_attack]type=fire which matches with fire type attacks even if the attack is modified by [damage_type]alternative_type=fire provided that the target is more sensitive to fire than to the original type of the attack

Comment (newfrenchy83) [reply src/formula/callable_objects.cpp:181]: @Pentarctagon I think this line is essential for the proper functioning of the code, as for the other attributes of the attack

File: src/formula/callable_objects.cpp (hunk 9 above)
Comment (newfrenchy83) [reply src/formula/callable_objects.cpp:93]: I put the last commit back in place

**Comment by CelticMinstrel** [comment]
(continuing from https://github.com/wesnoth/wesnoth/pull/8936#discussion_r1885638610 )

> I hesitate because of the infinite recursion if the formula is used from a [damage_type] affecting the checked attack, but if I modify the code of the formula, it is good for the attributes of filter_weapon and the formula to be identical. So, do I modify formula or not?

My concern was that formulas using `type` will change meaning, but it seems that is in fact the entire purpose of this PR (though in filters, with formulas just being an auxiliary case).

So, let's keep your change to the meaning of `type` in formula, but make two other changes:

* Instead of `original_type`, let's call it `base_type`.
* Can we add a key that returns a list of all the applicable types? It would only have multiple elements if `alternative_type` was used, otherwise it would be a single element that's equal to `type`.

-----

Now, we also need to consider the Lua side of things. As I understand, you found yourself unable to add attributes to the Lua attack object because these depend on the owning unit; and, unlike the case with filters, nothing enforces that a "specials context" is in effect when accessing the attack from Lua.

So, why don't we add a function, `wesnoth.units.effective_damage_type`? It could have the following signature (formatted according to the wiki documentation):

* **wesnoth.units.effective_damage_type**(_unit_, _weapon_) &rarr; _effective_damage_type_, [_alternative_damage_type_]

The "weapon" could be either:
* An index (starting from 1), selecting a weapon the unit owns. Error if they don't have that many weapons.
* The ID (ie, `name=` key) of a weapon the unit owns. Error if they don't own that weapon.
* A Lua attack object. Error if it's a weapon that belongs to a different unit or no unit.

**Comment by newfrenchy83** [comment]
> Now, we also need to consider the Lua side of things. As I understand, you found yourself unable to add attributes to the Lua attack object because these depend on the owning unit; and, unlike the case with filters, nothing enforces that a "specials context" is in effect when accessing the attack from Lua.
> 
> So, why don't we add a function, `wesnoth.units.effective_damage_type`? It could have the following signature (formatted according to the wiki documentation):
> 
>     * **wesnoth.units.effective_damage_type**(_unit_, _weapon_) → _effective_damage_type_, [_alternative_damage_type_]
> 
> 
> The "weapon" could be either:
> 
>     * An index (starting from 1), selecting a weapon the unit owns. Error if they don't have that many weapons.
> 
>     * The ID (ie, `name=` key) of a weapon the unit owns. Error if they don't own that weapon.
> 
>     * A Lua attack object. Error if it's a weapon that belongs to a different unit or no unit.

I prefer that this be the subject of another PR, because I don't know where to start, and this one has been open since May 30.

**Comment by CelticMinstrel** [comment]
> I prefer that this be the subject of another PR, because I don't know where to start, and this one has been open since May 30.

I would prefer it in the same PR, but I can accept it in a separate one.

(continuing from https://github.com/wesnoth/wesnoth/pull/8936/files#r1885648676 )

> This check is the code to check the type of damage actually inflicted while the old code for type() is reassigned to 'original_type', in short if you use 'type' in a formula in [filter_attack/weapon], it is the type of damage that will actually be used that will be detected, a bit like if you use [filter_attack]type=fire which matches with fire type attacks even if the attack is modified by [damage_type]alternative_type=fire provided that the target is more sensitive to fire than to the original type of the attack
> I think this line is essential for the proper functioning of the code, as for the other attributes of the attack

Are you misunderstanding something…? You're modifying `attack_type_callable::do_compare` here, which determines the sorting order of the attack. What does that have to do with the filter? I don't think filters do anything like sorting attacks. Furthermore, having the sorting order of the attack change based on map conditions seems like an absolutely terrible idea, so I think this check should be removed.

**Comment by newfrenchy83** [comment]
> Are you misunderstanding something…? You're modifying `attack_type_callable::do_compare` here, which determines the sorting order of the attack. What does that have to do with the filter? I don't think filters do anything like sorting attacks. Furthermore, having the sorting order of the attack change based on map conditions seems like an absolutely terrible idea (it will upset the balanced binary tree inside `std::map<wfl::variant, wfl::variant>`), so I think this check should be removed.

line removed.

File: src/formula/callable_objects.cpp (hunk 10 above)
Comment (newfrenchy83) [reply src/formula/callable_objects.cpp]: Line removed because unnecessary.

**Comment by newfrenchy83** [comment]
@Pentarctagon @CelticMinstrel Does anyone have any other comments to leave or can we consider merging this PR?

**Comment by CelticMinstrel** [comment]
It looks like this comment has not been addressed: https://github.com/wesnoth/wesnoth/pull/8936#issuecomment-2543923606

In particular, I asked to change `original_type` to `base_type`, but it looks like the diff still says `original_type`. You instead added `base_type` as a key in filters, which is fine, but not what I asked for. Please change the name in formulas to `base_type` too.

**Comment by newfrenchy83** [comment]
> It looks like this comment has not been addressed: [#8936 (comment)](https://github.com/wesnoth/wesnoth/pull/8936#issuecomment-2543923606)
> 
> In particular, I asked to change `original_type` to `base_type`, but it looks like the diff still says `original_type`. You instead added `base_type` as a key in filters, which is fine, but not what I asked for. Please change the name in formulas to `base_type` too.

'base_type' added in formulaas, sorrry i don't see a line.

**Comment by newfrenchy83** [comment]
> So, why don't we add a function, `wesnoth.units.effective_damage_type`? It could have the following signature (formatted according to the wiki documentation):
> 
>     * **wesnoth.units.effective_damage_type**(_unit_, _weapon_) → _effective_damage_type_, [_alternative_damage_type_]
> 
> 
> The "weapon" could be either:
> 
>     * An index (starting from 1), selecting a weapon the unit owns. Error if they don't have that many weapons.
> 
>     * The ID (ie, `name=` key) of a weapon the unit owns. Error if they don't own that weapon.
> 
>     * A Lua attack object. Error if it's a weapon that belongs to a different unit or no unit.

I'm not sure that trying to add this to lua at all costs is such a good idea, in lua_unit_attack, the damage and num_attacks() attributes refer to the values ​​encoded in [attack] without taking into account modifications by special weapons or [leadership], partly because the activity of the special weapon used cannot be checked (if any filter is used, it will return a false value); and everyone accepts that. It should be the same for the damage type, and your idea of ​​a function will solve nothing or very partially (the opponent's data and its resistance will still be unavailable), I will also remove the modifications I have already made

**Comment by Pentarctagon** [comment]
@newfrenchy83 have celticminstrel's comments been addressed then?

**Comment by newfrenchy83** [comment]
> @newfrenchy83 have celticminstrel's comments been addressed then?

Like unit information and oppoent can't be passed in attack lua table, i cannot satisfy comment of @CelticMinstrel  on this point. but i fixed formula

**Comment by newfrenchy83** [comment]
@Pentarctagon @stevecotton Could you merge this PR, since @CelticMinstrel is asking me for something unrealizable regarding lua

**Comment by newfrenchy83** [comment]
@Pentarctagon thank you.
